app/xfwp/xfwp.man
.\" $Xorg: xfwp.man,v 1.4 2001/02/09 02:05:46 xorgcvs Exp $
.\" Copyright 1996, 1998  The Open Group
.\"
.\" Permission to use, copy, modify, distribute, and sell this software and its
.\" documentation for any purpose is hereby granted without fee, provided that
.\" the above copyright notice appear in all copies and that both that
.\" copyright notice and this permission notice appear in supporting
.\" documentation.
.\"
.\" The above copyright notice and this permission notice shall be included
.\" in all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
.\" IN NO EVENT SHALL THE OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR
.\" OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
.\" ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
.\" OTHER DEALINGS IN THE SOFTWARE.
.\"
.\" Except as contained in this notice, the name of The Open Group shall
.\" not be used in advertising or otherwise to promote the sale, use or
.\" other dealings in this Software without prior written authorization
.\" from The Open Group.
.\"
.\"
.\" $XFree86: xc/programs/xfwp/xfwp.man,v 1.7 2001/02/07 22:35:23 tsi Exp $
.\"
.nh
.TH XFWP 1 __xorgversion__
.SH NAME
xfwp - X firewall proxy
.SH SYNOPSIS
.B xfwp
[option ...]
.PP
.SH COMMAND LINE OPTIONS
The command line options that can be specified are:
.PP
.TP 8
.B \-cdt \fInum_secs\fP
Used to override the default time-to-close (604800 seconds) for xfwp client
data connections on which there is no activity (connections over which
X protocol is already being relayed by xfwp)
.PP
.TP 8
.B \-clt \fInum_secs\fP
Used to override the default time-to-close (86400 seconds) for xfwp client
listen ports (ports on xfwp to which X clients first connect when trying to
reach an X server)
.PP
.TP 8
.B \-pdt \fInum_secs\fP
Used to override the default time-to-close (3600 seconds) for Proxy Manager
connections on which there is no activity
.PP
.TP 8
.B \-config \fIfile_name\fP
Used to specify the configuration the name of the configuration file
.PP
.TP 8
.B \-pmport \fIport_number\fP
Used to override the default port address (4444) for proxy manager connections
.PP
.TP 8
.B \-verify
Used to display the configuration file rule that was actually matched for
each service request
.PP
.TP 8
.B \-logfile \fIfile_name\fP
Used to specify the name of a file where audit information should be logged.
The format of a logged entry is: time of day; event code; source IP address;
destination IP address; and configuration rule number.  The event codes
are: "0" for a successful connection; "1" if a connection is denied because of
a configuration rule; and "2" if a connection is denied because of an
authorization failure.  If the event code is "1", and a configuration file
is used, the configuration rule number is the line number of the
configuration file where the match was made (see the section
CONFIGURATION FILE for more information).  If the event code is not "1",
or if no configuration file is used, the configuration rule number is "-1".
.PP
.TP 8
.B \-loglevel \fI{0,1}\fP
Used to specify the amount of audit detail that should be logged.  If "0",
all connections are logged.  If "1", only unsuccessful connections are logged.
.PP
.TP 8
.B \-max_pm_conns \fInum_connections\fP
Used to specify the maximum number of Proxy Manager connections.  The
default is 10.
.PP
.TP 8
.B \-max_pm_conns \fInum_connections\fP
Used to specify the maximum number of X server connections.  The
default is 100.
.PP
.SH DESCRIPTION
The X firewall proxy (xfwp) is an application layer gateway proxy
that may be run on a network firewall host to forward X traffic
across the firewall.  Used in conjunction with the X server Security
extension and authorization checking, xfwp constitutes a safe, simple,
and reliable mechanism both to hide the addresses of X servers located
on the Intranet and to enforce a server connection policy.  Xfwp cannot
protect against mischief originating on the Intranet; however, when
properly configured it can guarantee that only trusted clients originating
on authorized external Internet hosts will be allowed inbound access to
local X servers.

To use xfwp there must be an X proxy manager running in the local environment
which has been configured at start-up to know the location of the xfwp.
[NOTE:  There may be more than one xfwp running in a local environment;
see notes below on load balancing for further discussion.]  Using the
xfindproxy utility (which relays its requests through the proxy manager)
a user asks xfwp to allocate a client listen port for a particular X server,
which is internally associated with all future connection requests for that
server.  This client listen port address is returned by the proxy manager
through xfindproxy.  The xfwp hostname and port number is then passed
out-of-band (i.e., via a Web browser) to some remote X client, which will
subsequently connect to xfwp instead of to the target X server.

When an X client connection request appears on one of xfwp's listen ports,
xfwp connects to the X server associated with this listen port and performs
authorization checks against the server as well as against its own configurable
access control list for requesting clients.  If these checks fail, or if
the requested server does not support the X Security Extension, the client
connection is refused.  Otherwise, the connection is accepted and all ensuing
data between client and server is relayed by xfwp until the client terminates
the connection or, in the case of an inactive client, until a configured
timeout period is exceeded.  Xfwp is designed to block while waiting for
activity on its connections, thereby minimizing demand for system cycles.

If xfwp is run without a configuration file and thus no sitepolicy is
defined, if xfwp is using an X server where xhost + has been run to turn
off host-based authorization checks, when a client tries to connect to
this X server via xfwp, the X server will deny the connection.  If xfwp
does not define a sitepolicy, host-based authorization must be turned on
for clients to connect to an X server via the xfwp.
.PP
.SH INTEROPERATION WITH IP PACKET-FILTERING ROUTERS
The whole purpose of the xfwp is to provide reliable control over access
to Intranet X servers by clients originating outside the firewall.  At
the present time, such access control is typically achieved by firewall
configurations incorporating IP packet-filtering routers.  Frequently,
the rules for such filters deny access to X server ports (range 6000 -
6xxx) for all Intranet host machines.

In order for xfwp to do its job, restrictions on access for ports 6001 - 6xxx
must be removed from the rule-base of the IP packet-filtering router.  [NOTE:
xfwp only assigns ports in the range beginning with 6001; access to port
6000 on all Intranet hosts may continue to be denied.]  This does not
mean the Intranet firewall will be opened for indiscriminate entry by X
clients.  Instead, xfwp supports a fully configurable rule-based access
control system, similar to that of the IP packet-filter router itself.
Xfwp in effect adds another level of packet-filtering control which is
fully configurable and applies specifically to X traffic.  See section
entitled CONFIGURATION FILE, below, for further details.
.PP
.SH INSTALLATION, SETUP AND TROUBLESHOOTING
Xfwp is typically run as a background process on the Intranet firewall host.
It can be launched using any of the command-line options described above.
As noted above, xfwp works only in conjunction with proxy manager and the
xfindproxy utility.  It can also be configured to support a user-defined
X server site security policy, in which the X server is required to indicate
to xfwp whether or not it supports the particular policy.  Consult the
X server man pages for further information on these components.  Xfwp
diagnostics can be turned on by compiling with the -DDEBUG switch.
Connection status can be recorded by using the -logfile and -loglevel
command line options.
.PP
.SH PERFORMANCE, LOAD BALANCING AND RESOURCE MANAGEMENT
Xfwp manages four different kinds of connections:  proxy manager (PM) data,
X client listen, X client data, and X server.  The sysadmin employing xfwp
must understand how the resources for each of these connection types are
allocated and reclaimed by xfwp in order to optimize the availability of
xfwp service.

Each connection-type has a default number of allocation slots and
a default timeout.  The number of allocation slots for PM connections
and X server connections is configurable via command line options.
Connection timeouts are also configurable via command line options.
Each connection timeout represents the period the connection
will be allowed to remain open in the absence of any activity on that
connection.  Whenever there is activity on a connection, the time-to-close
is automatically reset.  The default distribution of total process connection
slots across the four connection types, as well as the choice of default
timeouts for the connection types, is governed by a number of assumptions
embedded in the xfwp use model.


The default number of PM connections is 10 and the
default duration for PM connections is 3,600
seconds (1 hour) for each connection after time of last activity.
At start-up, xfwp listens for PM connection requests on any non-reserved
port (default of 4444 if not specified on the xfwp command-line).  The PM
normally connects to xfwp only when a call is made to the PM with xfindproxy.
Thereafter, the PM remains connected to xfwp, even after the messaging between
them has been completed, for the default connection duration period.  In some
cases this may result in depletion of available PM connection slots.
If the sysadmin expects connections to a single xfwp from many PM's,
xfwp should be started using the -pdt command line option, with a timeout
value reflecting the desired duration that inactive connections will be
permitted to remain open.

Xfwp client listeners are set up by a call to xfindproxy and continue to
listen for X client connection requests for a default duration of 86,400
seconds (24 hours) from the point of last activity.  After this time they
are automatically closed and their fd's recovered for future allocation.
In addressing the question of how to choose some alternative timeout
value which will guarantee the availability of client listen ports,
sysadmins should take into consideration the expected delay between
the time when the listener was allocated (using xfindproxy) and the time
when a client actually attempts to connect to xfwp, as well the likelihood
that client listeners will be re-used after the initial client data
connection is closed.

Each client connection is allocated a default lifetime of 604,800
seconds (7 * 24 hours)
from the point when it last saw activity.  After this time it is
automatically closed and its fd's recovered for future allocation.
Because server connections are not actually established until a connection
request from a remote X client arrives at one of the xfwp's client listen
ports, the client data timeout applies both to client-xfwp connections as well
as to xfwp-server connections.  If the system administrator expects many
client data connections through xfwp, an overriding of the default timeout
should be considered.
.PP
.SH CONFIGURATION FILE
The xfwp configuration file resides on the xfwp host machine and is
used to determine whether X client data connection requests will be
permitted or denied.  The path to the file is specified at start-up
time.  If no configuration file is specified, all X client data
connection requests routed through xfwp will be by default permitted,
assuming that other X server authorization checks are successful.  If
a configuration file is supplied but none of its entries matches the
connection request then the connection is by default denied.

If a line in the configuration file begins with the '#' character
or a new-line character, the line is ignored and the evaluator will
skip the line.

The configuration file supports two entirely independent authorization
checks:  one which is performed by xfwp itself, and a second which is the
result of xfwp's querying the target X server.  For the first of these,
the configuration file employs a syntax and semantic similar to that of IP
packet-filtering routers.  It contains zero or more source-destination
rules of the following form:
.PP
{permit | deny} <src> <src mask> [<dest> <dest mask> [<operator> <service>]]
.sp
.IP permit/deny 12
the keywords ``permit'' or ``deny'' indicate whether the
rule will enable or disable access, respectively
.IP src 12
the IP address against the host who originated the
connection request will be matched, expressed in IP
format (x.x.x.x)
.IP "src mask" 12
a subnet mask, also in IP format, for further qualifying
the source mask.  Bits set in the mask indicate bits of the
incoming address to be \fIignored\fP when comparing to the specified src
.IP dest 12
the IP address against which the destination of the
incoming connection request (i.e. the host IP of the
X server to which the incoming client is attempting to
connect) will be matched
.IP "dest mask" 12
a subnet mask, also in IP format, for further qualifying
the destination mask.  Bits set in the mask indicate bits of the
destination address to be \fIignored\fP when comparing to the specified dest
.IP operator 12
always ``eq'' (if the service field is not NULL)
.IP service 12
one of the following three strings:  ``pm'', ``fp'', or
``cd'', corresponding to proxy manager, xfindproxy, or
client data, respectively
.PP
For the second type of authorization check, the configuration file contains
zero or more site policy rules of the following form:
.PP
{require | disallow} sitepolicy <site_policy>
.sp
.IP require 12
specifies that the X server \fImust\fP be configured with \fIat least one\fP
of the corresponding site policies, else it must refuse the connection.
.IP disallow 12
specifies that the X server \fImust not\fP be configured with \fIany\fP of
the corresponding site policies, else it must refuse the connection.
.IP sitepolicy 12
a required keyword
.IP "<site_policy>" 12
specifies the policy string.  The string may contain any
combination of alphanumeric characters subject
only to interpretation by the target X server
.PP
.SH RULES FOR EVALUATING THE XFWP CONFIGURATION FILE ENTRIES
For the first type of configurable authorization checking, access
can be permitted or denied for each connection type based upon
source and, optionally, destination and service.  Each file entry must
at a minimum specify the keyword ``permit'' or ``deny'' and the two
source fields.  The
destination and service fields can be used to provide finer-grained
access control if desired.
.PP
The algorithm for rule-matching is as follows:
.PP
.RS 3
  while (more entries to check)
  {
    if ((<originator IP> AND (NOT <src mask>)) == src)
      [if ((<dest X server IP> AND (NOT <dest mask>)) == dest)]
        [if (service fields present and matching)]
          do either permit or deny connection depending on keyword
    else
      continue
  }
  if (no rule matches)
    deny connection
.RE
.PP
If a permit or deny rule does not specify a service and operation, then
the rule applies to all services.  If a configuration file is specified
and it contains at least one valid deny or permit rule, then a host
that is not explicitly permitted will be denied a connection.
.PP
Site policy configuration checking constitutes a separate (and X server
only) authorization check on incoming connection requests.  Any number of
require or disallow rules may be specified, but all rules must be of the
same type; that is, a single rule file cannot have both ``require'' and
``disallow'' keywords.  The algorithm for this check is as follows:
.PP
.RS 3
  if (X server recognizes any of the site policy strings)
    if (keyword == require)
      permit connection
    else
      deny connection
  else
    if (keyword == require)
      deny connection
    else
      permit connection
.RE
.PP
The site policy check is performed by xfwp only if the source-destination
rules permit the connection.
.PP
.SH
EXAMPLES
.PP
.sp
\fC
.nf
\&# if and only if server supports one of these policies then authorize
\&# connections, but still subject to applicable rule matches
\&#
require sitepolicy policy1
require sitepolicy policy2
\&#
\&# deny pm connections originating on 8.7.6.5 [NOTE:  If pm service
\&# is explicitly qualified, line must include destination fields as
\&# shown.]
\&#
deny  8.7.6.5  0.0.0.0  0.0.0.0  255.255.255.255  eq  pm
\&#
\&# permit xfindproxy X server connects to anywhere [NOTE:  If
\&# fp service is explicitly qualified, line must include source fields
\&# as shown.]
\&#
permit  0.0.0.0  255.255.255.255   0.0.0.0  255.255.255.255  eq  fp
\&#
\&# permit all connection types originating from the 192.0.0.0
\&# IP domain only
\&#
permit  192.0.0.0   0.255.255.255
.fi
\fP
.PP
Care should be taken that source-destination rules are written in the correct
order, as the first matching rule will be applied.  In addition to parser
syntax checking, a special command-line switch (-verify) has been provided
to assist the sysadmin in determining which rule was actually matched.
.PP
.SH BUGS
.PP
Xfwp should check server site policy and security extension before
allocating a listen port.
.PP
.SH SEE ALSO
xfindproxy (1), Proxy Management Protocol spec V1.0, proxymngr(1), Xserver(1)
.SH AUTHOR
Reed Augliere, consulting to X Consortium, Inc.


xserver/ChangeLog
commit e597f0119cd69b6d9edf86d06d941468f90d8e6d
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 27 21:07:17 2012 -0800

    configure.ac: 1.11.4
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit 6c491463ee7a3637c8aae7ddae9625b36b792078
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 20 23:54:58 2012 -0800

    1.11.3.902 (1.11.4 RC2)
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit b747a91485184649c05aba6fa10734f70e926032
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Wed Jan 18 11:47:21 2012 -0800

    XQuartz: Bump bundle version to 2.7.1
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit a1ed67fa0551ee2b39d002d3b786a2017b37c7ac
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Fri Jan 6 13:20:45 2012 +1000

    render: don't bother with animated cursors on floating slaves (#39989)
    
    X.Org Bug 39989 <http://bugs.freedesktop.org/show_bug.cgi?id=39989>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit bbb6b8c834e0e1491ca14403b5d0840dd14380d3)

commit f89c39e6c9e20ffc3c1ae228ee601c2ab9007ff4
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Mon Jan 9 15:04:10 2012 +1000

    dix: Update pointer limits for floating devices too (#43635)
    
    When the screen is restructured, the pointer limits need to be reset for
    floating slave devices as well, not just for master pointers. Only skip
    devices that don't have a cursor (attached slaves and keyboard)
    
    Bug reproducer: float an absolute slave device, rotate the screen - the
    device is now confined to a section of the screen only.
    
    X.Org Bug 43635 <http://bugs.freedesktop.org/show_bug.cgi?id=43635>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Chase Douglas <chase.douglas@canonical.com>
    (cherry picked from commit 35bd77e9d0701daae87d681900d749604fc6471f)

commit dea9377a257a402b2c7fe4a96c958d951a7d4579
Author: Zhigang Gong <zhigang.gong@linux.intel.com>
Date:   Wed Jan 4 07:01:19 2012 +0000

    mi/mibitblt: Fix an overflow bug of bit shift.
    
    When depth equal to 32 and planeMask equal to 0, the overflow will
    occur and cause the pixmap can't be cleared. There are some test
    cases in XTS hit this bug, and this fix can eliminate the corresponding
    failures.
    
    Signed-off-by: Zhigang Gong <zhigang.gong@linux.intel.com>
    Reviewed-by: Adam Jackson <ajax@redhat.com>
    Signed-off-by: Keith Packard <keithp@keithp.com>
    (cherry picked from commit f4956faab9ccf9aba6cf9603f4489f5dad19a347)

commit d2db5100b1bafeeea8c1febd390631c8f8abb094
Author: Alan Coopersmith <alan.coopersmith@oracle.com>
Date:   Fri Oct 28 21:29:50 2011 -0700

    Convert DetermineClientCmd to use strdup instead of malloc+strncpy
    
    *cmdname is initialized to NULL earlier in the function, so it's
    okay to overwrite it with NULL if strdup fails, don't need that
    extra check.
    
    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit 780133f9ae7fada462714b47e79d26075bbd9abe)

commit 5b28bcb340e279e7f4e76617e86b439092367690
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:23:12 2011 -0500

    dix: Tune dixLookupDrawable for success
    
    The vast vast vast majority of resource lookups are successful.  Move some
    work to the error paths so we don't punish success.
    
    Before:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1148346.9/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2091666.1/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 8f9bdfd293ad8e45755efe8d764b4dcc2a724f51)

commit e1ea93fec93b13d18df4451210aaaf322b769b6a
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 20:23:40 2011 -0500

    dix: Tune dtrace hooks around Dispatch
    
    Don't call LookupMajorName if the hooks aren't active, it's quite expensive.
    
    Before:
    40000000 trep @   0.0009 msec (1087458.5/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2012238.6/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 83a98543b58c661a22574a6f8d8f9d777c0955b8)

commit 9eaa0b5c89f645c50ea90e07e8b49ebacf4349a0
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:00:05 2011 -0500

    xace: ricer tuning for XaceHook
    
    gcc doesn't want to hoist the check for XaceHooks[hook] != NULL above the
    varargs code for some reason, so do it ourselves.
    
    Before:
    40000000 trep @   0.0010 msec (1050420.2/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (1921147.6/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1087458.5/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2012238.6/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 11977fab546da35d579ebe285e3c26864007805e)

commit a3b6ecc893e17028cfbe16d8d7981c9037fc0766
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Tue Jan 3 10:45:45 2012 +1000

    xfree86: split warning about missing identifier or input driver
    
    Check for identifier first and bail if it's missing (also remove the current

app/xfwp/xfwp.c
/* $Xorg: xfwp.c,v 1.4 2001/02/09 02:05:45 xorgcvs Exp $ */

/*
Copyright 1996, 1998  The Open Group

Permission to use, copy, modify, distribute, and sell this software and its
documentation for any purpose is hereby granted without fee, provided that
the above copyright notice appear in all copies and that both that
copyright notice and this permission notice appear in supporting
documentation.

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of The Open Group shall
not be used in advertising or otherwise to promote the sale, use or
other dealings in this Software without prior written authorization
from The Open Group.
*/

#include <stdio.h>
#include <stdlib.h>
#include <X11/Xos.h>
#include <netinet/in.h>

#include <X11/ICE/ICElib.h>

#include "xfwp.h"
#include "io.h"
#include "misc.h"
#include "pm.h"
#include "transport.h"

/*
 * Global variables
 */
struct clientDataStruct 	global_data;  	/* for ICE callbacks */

struct pm_conn_buf 		** pm_conn_array;
struct server_list 		** server_array;
struct client_conn_buf 		** client_conn_array;

char **SitePolicies    = NULL;	/* list of site security policy strings */

int    SitePolicyCount = 0;	/* count of elements in SitePolicies */

Bool   SitePolicyPermit = 0;	/* True  := permit iff server supports
				 *      at least one listed policy,
			         *  False := deny if server has any of
				 *      the listed policies.
			         */

int
main (
    int 			argc,
    char 			* argv[])
{
    int 			* pm_listen_array;
    fd_set 			readable, writable, rinit, winit;
    int				nfds = 0;
    int 			nready = 0;
    struct ICE_setup_info 	pm_conn_setup;
    IceListenObj *		listen_objects;
    struct config *		config_info;

    /*
     * setup the global client data struct; we need to do this in order
     * to access program data in the ICE FWPprocessMessages() callback
     * without making everything global!  See FWPprotocolSetupProc() for
     * the rest of what we are doing
     */
    if ((config_info = malloc(sizeof(struct config))) == NULL)
    {
	fprintf(stderr, "Memory allocation failed, exiting\n");
	exit(1);
    }

    global_data.config_info = config_info;
    global_data.nfds =  &nfds;
    global_data.rinit = &rinit;
    global_data.winit = &winit;

    /*
     * complete the setup
     */
    doProcessInputArgs(config_info, argc, argv);

    if ((doHandleConfigFile(config_info)) == FAILURE)
	exit(1);

    if ((doInitDataStructs(config_info,
			   &pm_conn_setup)) == FAILURE)
	exit(1);

    /*
     * install the ICE i/o error handler
     */
    doInstallIOErrorHandler();

    /*
     * zero the select() read/write masks
     */
    FD_ZERO(&readable);
    FD_ZERO(&writable);
    FD_ZERO(&rinit);
    FD_ZERO(&winit);

    /*
     * create listener socket(s) for PM connections
     */
    if (!doSetupPMListen(config_info->pm_listen_port,
			 &config_info->num_pm_listen_ports,
			 &pm_listen_array,
			 &listen_objects,
			 &nfds,
			 &rinit))
	exit(1);

    while(1)
    {
	readable = rinit;
	writable = winit;

	doSelect (config_info,
		  &nfds,
		  &nready,
		  &readable,
		  &writable);

	doCheckTimeouts (config_info,
			 &nready,
			 &rinit,
			 &winit,
			 &readable,
			 &writable);

	doProcessSelect (&nfds,
			 &nready,
			 &readable,
			 &writable,
			 &rinit,
			 &winit,
			 pm_listen_array,
			 config_info,
			 &listen_objects);
    }
}
